clients/go/events/idempotent_recorder.go

```diff
@@ -0,0 +1,99 @@
+package events
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/golang/protobuf/proto"
+
+	eventsErr "github.com/lyft/flyteidl/clients/go/events/errors"
+	"github.com/lyft/flyteidl/gen/pb-go/flyteidl/core"
+	"github.com/lyft/flyteidl/gen/pb-go/flyteidl/event"
+
+	"github.com/lyft/flytestdlib/logger"
+	"github.com/lyft/flytestdlib/promutils"
+)
+
+type EmittedEventInfo struct {
+	lastUpdatedAt time.Time
+	phase         string
+}
+
+type idempotentWorkFlowEventRecorder struct {
+	internalRecorder    *eventRecorder
+	events              map[string]EmittedEventInfo
+	maxUpdateLagSeconds int64
+	mutex               *sync.Mutex
+}
+
+func (i *idempotentWorkFlowEventRecorder) idempotentRecord(ctx context.Context, id string, e proto.Message, phase string, terminal bool) error {
+	i.mutex.Lock()
+	defer i.mutex.Unlock()
+
+	existing, ok := i.events[id]
+	emitTime := time.Now()
+	if !ok || existing.phase != phase || emitTime.Sub(existing.lastUpdatedAt) > time.Duration(i.maxUpdateLagSeconds)*time.Second {
+		err := i.internalRecorder.sinkEvent(ctx, e)
+		if err != nil {
+			if !eventsErr.IsAlreadyExists(err) {
+				return err
+			}
+			logger.Infof(ctx, "Workflow event phase: %s, executionId %s already exist", phase, id)
+		}
+
+		// delete the event from events map if it is an terminal event
+		if ok && terminal {
+			delete(i.events, id)
```

Comment:
but then what if we send multiple terminal events? I still feel rather than having our own global lock, i think i prefer LRU cache? it also allows the map to now grow unbounded?

Reply:
terminal events will be emitted every time without going through the cache (first terminal event will remove the entry from cache).

OK, I  add a LRU cache.